### PR TITLE
Add ReplenishingRateLimiter and IdleDuration

### PR DESCRIPTION
**Pull request by BrennanConroy**
Part of https://github.com/dotnet/runtime/issues/65400#issuecomment-1075473959

### CODE REVIEW

File: src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ConcurrencyLimiter.cs [hunk 1]
```diff
@@ -23,10 +24,14 @@ public sealed class ConcurrencyLimiter : RateLimiter
         private static readonly ConcurrencyLease SuccessfulLease = new ConcurrencyLease(true, null, 0);
         private static readonly ConcurrencyLease FailedLease = new ConcurrencyLease(false, null, 0);
         private static readonly ConcurrencyLease QueueLimitLease = new ConcurrencyLease(false, null, 0, "Queue limit reached");
+        private static readonly double TickFrequency = (double)TimeSpan.TicksPerSecond / Stopwatch.Frequency;
 
         // Use the queue as the lock field so we don't need to allocate another object for a lock and have another field in the object
         private object Lock => _queue;
 
+        /// <inheritdoc />
+        public override TimeSpan? IdleDuration => _idleSince is null ? null : new TimeSpan((long)((Stopwatch.GetTimestamp() - _idleSince) * TickFrequency));
```
Comment (halter73) [inline src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ConcurrencyLimiter.cs:33]:
```suggestion
        public override TimeSpan? IdleDuration => _idleSince is null ? null : Stopwatch.GetElapsedTime(_idleSince);
```

File: src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ConcurrencyLimiter.cs [hunk 2]
```diff
@@ -213,6 +219,12 @@ private void Release(int releaseCount)
                         break;
                     }
                 }
+
+                if (_permitCount == _options.PermitLimit && _idleSince is null)
```
Comment (halter73) [inline src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ConcurrencyLimiter.cs:223]: Is it possible for `_idleSince` to be non-null here?

File: src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/TokenBucketRateLimiter.cs [hunk 3]
```diff
@@ -292,6 +304,12 @@ private void ReplenishInternal(uint nowTicks)
                         break;
                     }
                 }
+
+                if (_tokenCount == _options.TokenLimit && _idleSince is null)
```
Comment (halter73) [inline src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/TokenBucketRateLimiter.cs:308]: We could also just assert `_idleSince` is null here, right?

File: src/libraries/System.Threading.RateLimiting/tests/TokenBucketRateLimiterTests.cs [hunk 4]
```diff
@@ -744,5 +744,35 @@ public async Task ReplenishWorksWhenTicksWrap()
...
+
+        [Fact]
+        public override void IdleDurationUpdatesWhenChangingFromActive()
+        {
+            var limiter = new TokenBucketRateLimiter(new TokenBucketRateLimiterOptions(1, QueueProcessingOrder.OldestFirst, 2,
+                TimeSpan.Zero, 1, autoReplenishment: false));
+            limiter.Acquire(1);
+            limiter.TryReplenish();
+            Assert.NotNull(limiter.IdleDuration);
+        }
```
Comment (halter73) [inline src/libraries/System.Threading.RateLimiting/tests/TokenBucketRateLimiterTests.cs:776]: Do we have any tests verifying the new `ReplenishingRateLimiter` properties are what we expect?

File: src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/TokenBucketRateLimiter.cs [hunk 5]
```diff
@@ -11,11 +11,12 @@ namespace System.Threading.RateLimiting
     /// <summary>
     /// <see cref="RateLimiter"/> implementation that replenishes tokens periodically instead of via a release mechanism.
     /// </summary>
-    public sealed class TokenBucketRateLimiter : RateLimiter
+    public sealed class TokenBucketRateLimiter : ReplenishingRateLimiter
     {
         private int _tokenCount;
         private int _queueCount;
         private uint _lastReplenishmentTick = (uint)Environment.TickCount;
```
Comment (halter73) [inline src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/TokenBucketRateLimiter.cs:18]: Can we update this to use stopwatch too and get rid of the overflow logic?

File: src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ConcurrencyLimiter.cs (hunk 1 above)
Comment (BrennanConroy) [reply src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ConcurrencyLimiter.cs:33]: That API is new in 7.0 so can't use it

**Review by halter73 — APPROVED** [review]

**Pull request merged**